### PR TITLE
wip: refactor dst handling, add tests

### PR DESCRIPTION
**Pull request by nlwstein**
#### Summary of changes

**Asana Ticket:** [TICKET_NAME](TICKET_LINK)

[Please include a brief description of what was changed]

### CODE REVIEW

**Comment by nlwstein** [comment]
@jfabi CC: @bklebe @digitalcora This WIP branch should resolve the negative seconds DST issue that caused `commuter_rail_boarding` to crash over the weekend, but I am following an assumption that we want to keep the period the same (24h) on the edge case days vs shrinking / extending the period to keep `3am` as the cutover. 

My assumption is based on [this test](https://github.com/mbta/commuter_rail_boarding/blob/main/apps/commuter_rail_boarding/test/date_helpers_test.exs#L21), which seems to state this pretty explicitly (checking for time after 3a in DST-related test). 

However, [this thread](https://mbta.slack.com/archives/CF9QKK2AF/p1644433460521149) contradicts - Cora seems confident that it should be 3am. I can update this to follow that logic, but it will involve changing the existing tests. 

Do you have a definitive answer on when a service date rolls over on the DST changeover dates?

**Comment by nlwstein** [comment]
I have updated this branch to reflect the following: https://mbta.slack.com/archives/CF9QKK2AF/p1678825606437889?thread_ts=1678717417.225069&cid=CF9QKK2AF

File: apps/commuter_rail_boarding/test/date_helpers_test.exs [hunk 1]
```diff
@@ -28,12 +28,80 @@ defmodule DateHelpersTest do
...
+
+      # fall back
+      {:ambiguous, dt_one, dt_two} = local_dt(~N[2018-11-04T01:30:00])
+
+      for {seconds_until, local_dt} <- %{5400 => dt_one, 1800 => dt_two} do
+        {:ok, utc_datetime} = DateTime.shift_zone(local_dt, "Etc/UTC")
+
+        assert seconds_until =
+                 seconds_until_next_service_date(service_date(utc_datetime), utc_datetime)
+      end
```
Comment (bklebe) [inline apps/commuter_rail_boarding/test/date_helpers_test.exs:88]: question: I'm a bit confused by this part. Why are we testing the ambiguous date times? Only one of them is the correct one.

**Review by bklebe — COMMENTED** [review]
Overall this looks pretty good Nick, and I appreciate the comprehensive tests. However, I'm still uncomfortable with how much functionality is duplicated in https://github.com/mbta/commuter_rail_boarding/blob/main/apps/train_loc/lib/train_loc/utilities/time.ex, and also duplicated in tests for that module. What do you think about extracting this shared utility for calculating service dates into a new application in `apps/` that `train_loc` and `commuter_rail_boarding` both depend on?

File: apps/commuter_rail_boarding/test/date_helpers_test.exs (hunk 1 above)
Comment (nlwstein) [reply apps/commuter_rail_boarding/test/date_helpers_test.exs:88]: > question: I'm a bit confused by this part. Why are we testing the ambiguous date times? Only one of them is the correct one.

@bklebe This is copied logic from another test, but basically what it is doing is checking the seconds remaining are correct at both instances of 1:30am. So the seconds remaining will be different at both 1:30s (5400 before the timechange, 1800 the second time around) 🥴

File: apps/commuter_rail_boarding/test/date_helpers_test.exs [hunk 2]
```diff
@@ -28,12 +28,80 @@ defmodule DateHelpersTest do
...
+
+      # This is 30 minutes after the time going back, so 23.5 hours makes sense:
+      assert 1800 =
```
[the remark this answers is not in the document]
Comment (nlwstein) [reply apps/commuter_rail_boarding/test/date_helpers_test.exs:91]: > issue: 1800 seconds is 30 minutes, not 23.5 hours. 

I think I broke my comments when I changed the logic here, fixing...

File: apps/commuter_rail_boarding/test/date_helpers_test.exs [hunk 3]
```diff
@@ -28,12 +28,80 @@ defmodule DateHelpersTest do
...
+               )
+
+      assert 81_000 =
+               seconds_until_next_service_date(
+                 service_date(local_dt!(~N[2018-02-11T04:30:00])),
+                 local_dt!(~N[2018-02-11T04:30:00])
+               )
+    end
+
+    test "handles DST transitions" do
```
[the remark this answers is not in the document]
Comment (nlwstein) [reply apps/commuter_rail_boarding/test/date_helpers_test.exs]: Broke this one up.

**Comment by nlwstein** [comment]
> Overall this looks pretty good Nick, and I appreciate the comprehensive tests. However, I'm still uncomfortable with how much functionality is duplicated in https://github.com/mbta/commuter_rail_boarding/blob/main/apps/train_loc/lib/train_loc/utilities/time.ex, and also duplicated in tests for that module. What do you think about extracting this shared utility for calculating service dates into a new application in `apps/` that `train_loc` and `commuter_rail_boarding` both depend on?

@bklebe I have split out the `service date` logic out into a `shared` application that both applications now depend on. The `get_service_date` function has been whacked from `train_loc` and the references have been updated to the equivalent shared lib func.

**Review by bklebe — APPROVED** [review]
